### PR TITLE
🌱 [ci]: Skipping jobs which are not required when the changes are in only docs

### LINKED ISSUE #2591: [ci] - Stop to run all tests when changes are only in the docs
### What do you want to happen?

Currently, if we change a doc we still running all test
The goal of this task is to skip when possible the tests when the change is performed only in the MD files. 

### Extra Labels

_No response_

### PR DESCRIPTION
**Pull request by NikhilSharmaWe**
### Description
There is no need of running all jobs when the changes are only in docs. This PR skips the github actions which are not required in those cases.

**Examples :** 
PR making changes in go files runs all tests: https://github.com/NikhilSharmaWe/kubebuilder/pull/31
PR making changes in md files skip jobs which are not required: https://github.com/NikhilSharmaWe/kubebuilder/pull/30

### Motivation
Fixes Part of: #2591

### CODE REVIEW

**Comment by NikhilSharmaWe** [comment]
@joelanford Could you please review the changes made.

**Comment by NikhilSharmaWe** [comment]
@camilamacedo86 I think there is some issue here because tests are running in both cases (not e2e).
PR changing md file https://github.com/NikhilSharmaWe/kubebuilder/pull/8
PR changing go file https://github.com/NikhilSharmaWe/kubebuilder/pull/7

Could you please explain what is the issue here with the changes. Or there is some other isssue with repo settings.

File: .github/workflows/unit-tests.yml
Comment (NikhilSharmaWe) [inline .github/workflows/unit-tests.yml:63]: @camilamacedo86 Now the PR is working, I was missing this statement. Just wanted to ask should we remove statements 46-47 where we declared needs to `test`, since it was not used anywhere so I removed it, is it reasonable?

File: .github/workflows/unit-tests.yml
Comment (camilamacedo86) [inline .github/workflows/unit-tests.yml:47]: Why we remove this one? Could we revert?

File: test/check-docs-only.sh
Comment (camilamacedo86) [inline test/check-docs-only.sh:44]: Nit could you please give space at the end of the file?

File: .github/workflows/unit-tests.yml
Comment (NikhilSharmaWe) [reply .github/workflows/unit-tests.yml:47]: I thought we need `needs: check_docs_only` in the file and since `needs: - test` was not used, I thought we could remove it.

**Review by camilamacedo86 — COMMENTED** [review]
All sorted out.

File: test/check-docs-only.sh
Comment (camilamacedo86) [inline test/check-docs-only.sh:35]: we do not have changelog/ dir so that seems that we should not have this line

File: test/check-docs-only.sh
Comment (camilamacedo86) [inline test/check-docs-only.sh:34]: we do not have website/ dir as well. I shows that the dir correct for Kubebuilder in this case is the doc/ dir

File: test/check-docs-only.sh
Comment (camilamacedo86) [inline test/check-docs-only.sh:33]: Have we this file?

File: .github/workflows/apidiff.yml
Comment (camilamacedo86) [inline .github/workflows/apidiff.yml:24]: do we need to duplicate the code OR could it be centralized in one test only? Did we check that?

File: .github/workflows/apidiff.yml
Comment (camilamacedo86) [inline .github/workflows/apidiff.yml:7]: WDYT about we update the doc: https://github.com/kubernetes-sigs/kubebuilder/blob/master/CONTRIBUTING.md#where-the-ci-tests-are-configured bu adding a new section such as "Skipped jobs" and explaining what are the jobs which are or not skipped when we identify that the only docs changed?

File: test/check-docs-only.sh
Comment (NikhilSharmaWe) [reply test/check-docs-only.sh:33]: Yes

File: test/check-docs-only.sh
Comment (camilamacedo86) [inline test/check-docs-only.sh:36]: Should we not ignore here any change in the doc/* directory?

@NikhilSharmaWe

**Comment by NikhilSharmaWe** [comment]
@camilamacedo86 Is it fine now?

File: .github/workflows/apidiff.yml
Comment (camilamacedo86) [reply .github/workflows/apidiff.yml:7]: @NikhilSharmaWe could we not:
1 - have a job just to check the skip with this code added here
2- make the others GitHub action only check its results

Why?
Then we have this logic centralized and not duplicated across all GitHub actions.

**Comment by NikhilSharmaWe** [comment]
If the approach used to prevent the duplication of the job `check_docs_only` is correct:

After the merge we can change we can change this statement https://github.com/kubernetes-sigs/kubebuilder/blob/647be49c440b42501d9eda63919e4062a98a7bf9/.github/workflows/apidiff.yml#L10
to `kubernetes-sigs/kubebuilder/.github/workflows/check_docs_only.yml@master`. Or should we do this before merge itself, please inform.

File: .github/workflows/apidiff.yml
Comment (camilamacedo86) [inline .github/workflows/apidiff.yml:10]: We cannot merge it with `NikhilSharmaWe/kubebuilder`
Also, it shows a path valid only in your local env/machine

**Review by camilamacedo86 — CHANGES_REQUESTED** [review]
We need to fix the path

**Comment by NikhilSharmaWe** [comment]
@camilamacedo86 I used `NikhilSharmaWe/kubebuilder/.github/workflows/check_docs_only.yml@checkDocsOnly` because it will show an error if I used kubernetes-sigs master (since there is no workflow check_docs_only.yml present right now). 

Now I have changed it to `kubernetes-sigs/kubebuilder/.github/workflows/check_docs_only.yml@master`, it is showing an error but I think once we merge it, it will work fine.

Please inform me If I am unreasonable here.

File: .github/workflows/apidiff.yml
Comment (camilamacedo86) [inline .github/workflows/apidiff.yml:10]: we cannot move in this way. 
That would mean that it would not work in other branches. 

@NikhilSharmaWe 
/hold

File: .github/workflows/apidiff.yml
Comment (camilamacedo86) [reply .github/workflows/apidiff.yml:7]: We discussed that it could be an improvement in a follow up as well.

File: .github/workflows/apidiff.yml
Comment (camilamacedo86) [reply .github/workflows/apidiff.yml:40]: sorted

**Review by camilamacedo86 — APPROVED** [review]
The change on this PR will result in skipping the following jobs when only docs files were changed.
Screenshot 2022-04-22 at 09 02 22

The main follow-ups that I can see here show to be we begin to skip the e2e tests done in the prow when only doc files are changed. So that, if the change is only on the docs we do not need to run all e2e tests unnecessarily.

**Review by everettraven — APPROVED** [review]
/lgtm

**Review by rashmigottipati — APPROVED** [review]

**Pull request merged**
